vpd-manager/include/utility/dbus_utility.hpp

```diff
@@ -454,5 +458,128 @@ inline bool isBMCReady()
+    }
+    return l_rc;
+}
+
+/**
+ * @brief An API to disable BMC reboot guard
+ *
+ * This API disables BMC reboot guard. This API has an inbuilt re-try mechanism.
+ * If Disable Reboot Guard fails, this API attempts to Disable Reboot Guard for
+ * 3 more times at an interval of 1 second. The 3 re-tries are executed
+ * asynchronously. If Disable Reboot Guard still fails after retries, this API
+ * logs a Critical PEL.
+ *
+ * @return On success, returns 0, otherwise returns -1.
+ */
+inline int DisableRebootGuard() noexcept
+{
+    int l_rc{constants::FAILURE};
+
+    // A lambda which executes the DBus call to disable BMC reboot guard.
+    auto l_executeDisableRebootGuard = []() -> int {
+        int l_dBusCallRc{constants::FAILURE};
+        try
+        {
+            auto l_bus = sdbusplus::bus::new_default();
+            auto l_method = l_bus.new_method_call(
+                constants::systemdDBusService, constants::systemdDBusObjectPath,
+                constants::systemdDBusManagerInterface, "StartUnit");
+            l_method.append("reboot-guard-disable.service", "replace");
+            l_bus.call_noreply(l_method);
+            l_dBusCallRc = constants::SUCCESS;
+        }
+        catch (const sdbusplus::exception::SdBusError& l_ex)
+        {}
+        return l_dBusCallRc;
+    };
+
+    if (constants::FAILURE == l_executeDisableRebootGuard())
+    {
+        std::function<void()> l_retryDisableRebootGuard;
+
+        // A lambda which tries to disable BMC reboot guard for 3 times at an
+        // interval of 1 second.
+        l_retryDisableRebootGuard = [&]() {
+            constexpr int MAX_RETRIES{3};
+            static int l_numRetries{0};
+
+            if (l_numRetries < MAX_RETRIES)
+            {
+                logging::logMessage("Disable Reboot Guard retry : " +
+                                    std::to_string(l_numRetries));
+                l_numRetries++;
+                if (constants::FAILURE == l_executeDisableRebootGuard())
+                {
+                    // sleep for 1 second before next retry
+                    const std::chrono::seconds l_sleepTime{1};
+                    std::this_thread::sleep_for(l_sleepTime);
+                    l_retryDisableRebootGuard();
+                }
+                else
+                {
+                    l_numRetries = 0;
+                    l_rc = constants::SUCCESS;
+                }
+            }
+            else
+            {
+                // Failed to Disable Reboot Guard even after 3 retries.
+                // TODO: Log a Critical PEL
+                logging::logMessage("Failed to Disable Reboot Guard after " +
+                                    std::to_string(MAX_RETRIES) + " re-tries");
+                l_numRetries = 0;
+            }
+        };
+
+        try
+        {
+            // start asynchronous retry thread and return
+            const auto l_retryResult = std::async(std::launch::async,
```

Comment:
Why we are doing retries in thread?
What are we getting out of adding these complexity? 

Why not a simple loop to retry? This is an atomic call?

Reply:
> Why we are doing retries in thread?
What are we getting out of adding these complexity?

In worst case scenario, the Disable Reboot Guard re-try loop can take 3 seconds. If we do this synchronously, `Parser::updateVpdKeyword` API can take upto 3 seconds. `BackupAndRestore` calls this API, so in worst case, every keyword update call will take 3 seconds. This can happen during BMC boot up as well.

To avoid this, we can have the re-tries in an asynchronous thread and  `Parser::updateVpdKeyword` execution will continue to take same time as it does today.

Reply:
> This is an atomic call?

I do not see any documentation on whether `std::async()` is atomic. What is the concern here with respect to atomicity?